packages/common-components/src/Toasts/ToastContext.tsx

```diff
@@ -0,0 +1,196 @@
+    bottomRightContainer: {
+      bottom: 0,
+      right: 0
+    },
+    bottomRightAppearBox: {
+      animation: `$slideLeft ${animation.translate}ms`,
+      [breakpoints.down("sm")]: {
+        animation: `$slideUp ${animation.translate}ms`
+      }
+    },
+    bottomLeftContainer: {
+      bottom: 0,
+      left: 0
+    },
+    bottomLeftAppearBox: {
+      animation: `$slideRight ${animation.translate}ms`,
+      [breakpoints.down("sm")]: {
+        animation: `$slideUp ${animation.translate}ms`
+      }
+    },
+    "@keyframes slideRight": {
+      from: { transform: "translate(-100%, 0)" },
+      to: { transform: "translate(0, 0)" }
+    },
+    "@keyframes slideLeft": {
+      from: { transform: "translate(100%, 0)" },
+      to: { transform: "translate(0, 0)" }
+    },
+    "@keyframes slideUp": {
+      from: { transform: "translate(0, 100%)" },
+      to: { transform: "translate(0, 0)" }
+    },
+    "@keyframes slideDown": {
+      from: { transform: "translate(0, -100%)" },
+      to: { transform: "translate(0, 0)" }
+    }
+  })
+})
+
+type ToastContextProps = {
+  autoDismiss?: boolean
+  dismissTimeout?: number
+  defaultPosition?: ToastPosition
+  children: React.ReactNode | React.ReactNode[]
+}
+
+interface ToastContext {
+  addToast(toastParams: ToastParams): string
+  updateToast(toastId: string, toastParams: ToastParams, startDismissal?: boolean): void
+  removeToast(toastId: string): void
+  toasts: Toast[]
+}
+
+const ToastContext = React.createContext<ToastContext | undefined>(
+  undefined
+)
+
+const ToastProvider = ({
+  children,
+  autoDismiss = true,
+  defaultPosition = "topRight",
+  dismissTimeout = 5000
+}: ToastContextProps) => {
+  const classes = useStyles()
+  const [toasts, setToasts] = useState<Toast[]>([])
+
+  const removeToast = useCallback((toastId: string) => {
+    setToasts(toasts.filter((toast) => toast.id !== toastId))
+  }, [toasts])
+
+  const addToast = useCallback((toastParams: ToastParams) => {
+    const id = uuidv4()
+    setToasts((toasts) => ([
+      ...toasts,
+      { id,
+        ...toastParams,
+        toastPosition: toastParams.toastPosition || defaultPosition
+      }
+    ]))
+
+    const isProgressToast = toastParams.progress !== undefined
+    const shouldDismiss = toastParams.autoDismiss !== undefined ? toastParams.autoDismiss : autoDismiss
+    const dismissTimeOut = toastParams.dismissTimeout || dismissTimeout
+
+    if (shouldDismiss && !isProgressToast) {
+      setTimeout(() => {
+        setToasts((toasts) => toasts.filter((toast) => toast.id !== id))
+      }, dismissTimeOut)
+    }
+
+    return id
+  }, [autoDismiss, dismissTimeout, defaultPosition])
+
+  const updateToast = useCallback((toastId: string, toastParams: ToastParams, startDismissal?: boolean) => {
+    const dismissTimeOut = toastParams.dismissTimeout || dismissTimeout
+    if (startDismissal) {
+      setTimeout(() => {
+        setToasts((toasts) => toasts.filter((toast) => toast.id !== toastId))
```

Comment:
you can call `removeToast` here instead.
```suggestion
        removeToast(toast.id)
```

Reply:
Actually thats what I tried first!
But somehow the function keeps a reference of older toasts state from when it got initialized, and causes weird behaviour,
moved to this implementation later!

Reply:
ah ok interesting, the fact that `toast` is part of the dep array should prevent that. I'll take a closer look just for my information.

Reply:
@tanmoyAtb, would you mind telling me in what case this happened?, I can't reproduce the problem you mentioned, I've monitored the call to `removeToast` for instance when I delete several files, or upload several files (not at the same time), and all the toasts get removed in the right order.

Reply:
hey @Tbaut, so try starting two uploads parallely, 
both toasts would get dismissed whenever the first toasts gets dismissed!

Reply:
Yup that's what I did, and it's looking fine for me:

https://user-images.githubusercontent.com/33178835/132722698-7884b1e8-c5f2-4341-8c69-e341e1c830d5.mp4

Reply:
I think you might see the weird behaviour if you start uploads of similar size, where the startDismiss fires at close times!
In this recording, you can see toasts reappearing after uploads complete!

https://user-images.githubusercontent.com/10713353/132729395-a1c08678-6006-408c-868d-be178a3fad60.mov

Reply:
Thanks a lot, I got it, the sate that the `setTimeout` restores, is the one from a couple seconds ago, because of the closure. This is a very good example where we would actually need to have a `useRef` instead of `useState`. Allow me to hold this PR a couple more hours, for the sake of having behavior we 100% understand without duplication. I'll create a PR pointing to this one.

Reply:
Ok here we go, I've submitted here: https://github.com/ChainSafe/ui-monorepo/pull/1513
Let me know what you think.
For more info about the closure thing, you can have a look here for instance https://github.com/facebook/react/issues/14010